src/main/java/wiresegal/wob/arcanum/Parsing.kt

```diff
@@ -167,6 +168,57 @@ fun about(message: Message) {
     })
 }
 
+val progressCache = mutableListOf<Pair<String, String>>()
+var progressCacheTimeStamp: Long? = null
+
+fun showProgressBar(message: Message) {
+    val projects = getProgressBar()
+    var maxLength = 0
+    for (project in projects) {
+        maxLength = Integer.max(maxLength, project.first.length)
+    }
+    var progresses = ""
+    for (project in projects) {
+        val (name, percent) = project
+        // TODO: The whitespace doesn't show up in the message
+//        val wordWhiteSpace = " ".repeat(maxLength - name.length)
+//        val percentWhiteSpace = " ".repeat(Integer.max(0, 3 - percent.length))
+//        progresses += "$name$wordWhiteSpace is at $percentWhiteSpace$percent%\n"
+        progresses += "$name is at $percent%\n"
+    }
+
+    message.channel.sendMessage(EmbedBuilder().apply {
+        setTitle("Progress Bars")
+        setColor(embedColor)
+
+        setDescription(progresses)
+    })
+}
+
+fun getProgressBar(): MutableList<Pair<String, String>> {
+    if (progressCacheTimeStamp != null && (System.currentTimeMillis() - progressCacheTimeStamp!! <= progressCachePersistence)) {
+        return progressCache
+    } else {
+        val html = URL(homepageTarget).readText()
+        // This regex matches all <small> tags with class="vc_label", captures text with an optional <a> tag surrounding it, and then captures the numbers of the percentage inside the <span>
+        // I've added whitespace everywhere it can appear, but if Brandon decides to use exoteric characters in the future, this might break.
+        val regex = Regex("<small[\\s]*class=\"vc_label\"[\\s\\w:#;=\"]*>[\\s]*(?:<a[-\\s\\w:#;=\"/,.]*>)?([\\s\\w&.,:]*)(?:</a>)?[\\s]*<span[\\s\\w:#;=\"]*>[\\s]*([0-9]*)%[\\s]*</span>[\\s]*</small>")
```

Comment:
This seems *incredibly* difficult to read and maintain, not to mention fragile. I don't have a better idea at the moment, but I know that if I had to make changes to it, I would find it easier to rewrite the whole thing than to parse and update this regex (though I also hate regexes).

Another thing to consider is that this is very strongly coupled with the current design of Brandon's site. He doesn't change it all the time, but this is the fourth iteration in the last 15 or so years, and every version has used different way of defining the progress bars. For example, here's a not-very-well-formatted excerpt from the previous design:
```html
<span class="book-title">Stormlight 4 &amp; 5 outlining</span>
<div class="progress">
		<div class="wppb-wrapper after"><div class="after">80 %</div><div class="wppb-progress fixed"><span class="backimg" style="width: 88px; background: red;"><span></span></span></div></div>        </div>
```

So I think that, ideally, you would want a more dedicate parser, one where you can more easily specify what you are looking for in the HTML glob. Maybe something where you say "here's how to find the area where the progress bars are, and here's how to find all the labels, and here's how to find all the percentages" so that if one or more of those pieces has to change in the future, we don't have to overhaul the entire thing.

Reply:
This does seem like a big problem. Though the only thing that needs to change is the single line that implements the regex. It's quite difficult to make anything more readable than a boiler-plate code around regex like I've just done. It seems like it would need to be changed no matter what you do, without some very advanced context-based search through the text

Reply:
Well, this is at least more readable.

Reply:
The bot already uses the [Jsoup](https://jsoup.org/) HTML parsing library, which exposes a jQuery style selector API which makes dealing with DOM structures like this one very easy. That will make the code much more robust against slight future changes in the markup and especially more readable.

Reply:
I guess with this it would be possible to more quickly filter across the different tags. Wouldn't be future-proof, but should make it quite easy to maintain

Reply:
Jsoup certainly makes a lot more sense for searching through an html document. Hopefully this is quite a bit more readable.